contracts/tropykus/TropykusLendingService.sol

```diff
@@ -29,108 +26,56 @@ contract TropykusLendingService is LendingService, TropykusCommon {
         uint256 amount,
         uint256 listingId,
         address wallet
-    )
-        public
-        payable
-        override
-        withSubscription(mtx.req.from, listingId, wallet)
-    {
-        uint256 amountToLend;
-        ServiceListing memory listing = listings[listingId];
+    ) public payable override {
+        address currencyToLend = listings[listingId].currency;
+        uint256 _amount = msg.value > 0 ? msg.value : amount;
 
-        {
-            if (!listing.enabled) {
-                revert ListingDisabled(listingId);
-            }
-
-            amountToLend = amount;
-            if (listing.currency == address(0)) amountToLend = msg.value;
-            if (amountToLend == 0) {
-                revert InvalidAmount(amountToLend);
-            }
-
-            if (
-                listing.maxAmount < amountToLend ||
-                listing.minAmount > amountToLend
-            ) {
-                revert InvalidAmount(amountToLend);
-            }
+        if (_amount == 0) {
+            revert ZeroAmountNotAllowed({currency: currencyToLend});
         }
+        _onlyValidListingArgs(listingId, _amount);
+
+        emit Lend({
+            listingId: listingId,
+            lender: msg.sender,
+            currency: currencyToLend,
+            amount: msg.value
+        });
 
+        _removeLiquidityInternal(_amount, listingId);
+        _withSubscription(mtx.req.from, listingId, wallet);
```

Comment:
The idea of the modifier was to make it easy for developers to integrate it and easier for us to check if its implemented, what's the advantage of having it as a function? We can discuss this in a future refinement too

Reply:
Great question, I had to change it because slither was raising a high vulnerability on how that specific function gets called, if being executed within the modifier, a reentrancy issue can happen (according to slither), the first thing to be executed will be the modifier and several variables are updated after making that call. 

So I ensure `_withSubscription` gets called after changing state variables, which made move out the code from the modifier to a function that can be called after state changes are done.

I'm flexible to discuss if we should changed regardless of the the warning (though  I think we shouldn't).

Reply:
We can discuss it in the refinement and merge it for now, I'm just not sure about the vulnerability, its supposed to be one because we are doing an external call but that external call is our own smart contract and we can ensure that nothing bad is going to occur.

Reply:
Another idea that I have is just to move the logic from the start of the function to the end, that way we only call the contract at the end of the function. so instead of doing 

``` 
modifier  withSubscription {
   ...code...
   _;
}
```

we change it to:

``` 
modifier  withSubscription {
   _;
   ...code...
}
```

Reply:
Not really, the only concern is to be able to confidently and easily check that any protocol that extends the contract makes use of the function

Reply:
Ok cool, I implemented your suggestion, thanks!